app/admin/vaults/unlock-result.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/**
+ * Script to disaply vault edit result
+ *************************************/
+
+/* functions */
+require_once( dirname(__FILE__) . '/../../../functions/functions.php' );
+
+# initialize user object
+$Database       = new Database_PDO;
+$User           = new User ($Database);
+$Admin          = new Admin ($Database, false);
+$Result         = new Result ();
+$Password_check = new Password_check ();
+
+# verify that user is logged in
+$User->check_user_session();
+# check maintaneance mode
+$User->check_maintaneance_mode ();
+
+# make sure user has access
+if ($User->get_module_permissions ("vaults")==User::ACCESS_NONE) { $Result->show("danger", _("Insufficient privileges").".", true); }
+
+# strip input tags
+$_POST = $Admin->strip_input_tags($_POST);
+
+# validate csrf cookie
+$User->Crypto->csrf_cookie ("validate", "vaultunlock", $_POST['csrf_cookie']) === false ? $Result->show("danger", _("Invalid CSRF cookie"), true) : "";
+
+// fetch vault
+$vault = $Admin->fetch_object("vaults", "id", $_POST['vaultId']);
+// validate vault id
+$vault===false ? $Result->show("danger", _("Invalid ID"), true) : null;
+
+// test
+if($User->Crypto->decrypt($vault->test, $_POST["vaultpass"])!="test") {
+	$Result->show("danger", _("Invalid master password"), true);
+}
+else {
+	// write session
+	$_SESSION['vault'.$vault->id] = $_POST['vaultpass'];
```

Comment:
Hi @GaryAllan , currently key to decrypt is stored to user session in plain text, got better idea ?

Reply:
Hi Miha,
First thoughts....If stored in $_SESSION and config.php `$session_storage = "database";` the key is going to be written to the database unprotected. If `$session_storage = "files";` it's going to be written unprotected to a file somewhere, probably under /tmp.

I'll take a look over the next few days. Have you looked at MySQL/MariaDB at rest encryption (database manages the crypto) or client side JavaScript decryption (Browser manages the crypto)?

Reply:
I agree for sessions, there should be some better solution... MariaDB seems complicated for general use with encryption plugin etc.

Reply:
How about when a vault is opened we generate a random buffer ($key_a) and store this in $_SESSION and store the decryption key XOR'd with $key_a in a browser cookie set to expire with the browser is closed (cookie requires HTTPS).

If $key_a is random enough, the cookie and $_SESSION data are useless by themselves and can be combined when required for decryption (or use AES encryption instead of XOR, I need to think about it some more).

Anyhow....getting late. Speak tomorrow.

Reply:
Sure, no hurry, thanks !

Reply:
Hi Miha,
If there is only one vault it doesn't show the unlock option

![image](https://user-images.githubusercontent.com/18753294/94158409-424c8e00-fe7a-11ea-8e97-3346ff4fd64a.png)